src/Microsoft.OData.Core/UriParser/SemanticAst/ODataPath.cs

```diff
@@ -82,6 +82,14 @@ public int Count
             get { return this.segments.Count; }
         }
 
+        /// <summary>
+        /// Get the List of ODataPathSegments.
+        /// </summary>
+        public List<ODataPathSegment> List

```

Comment:
>List<ODataPathSegment> [](start = 15, length = 22)

In my opinion, to expose List<T> from a public class is not a good design.

1. List<T> exposes many more members than the IList<T>
2. If we change the type of private filed "segments" as other IEnumerable<T> implementation, what's the design goal about this public property?

Reply:
Would ReadOnlyCollection<T> give you the functionality you need, without allowing the user to change the collection (which could mess things up)?  You can call segments.AsReadOnly() the first time it's accessed to get the ReadOnlyCollection and then return the same instance in subsequent requests to avoid additional allocations.

---
In reply to: [352913193](https://github.com/OData/odata.net/pull/1603#discussion_r352913193) [](ancestors = 352913193)